spdx/v2/v2_2/document.go

```diff
@@ -98,27 +99,46 @@ func (d *Document) UnmarshalJSON(b []byte) error {
 
 	*d = Document(d2)
 
+	relationshipExists := map[string]bool{}
+	serializeRel := func(r *Relationship) string {
+		return fmt.Sprintf("%v-%v->%v", common.RenderDocElementID(r.RefA), r.Relationship, common.RenderDocElementID(r.RefB))
+	}
+
+	// index current list of relationships to ensure no duplication
+	for _, r := range d.Relationships {
+		relationshipExists[serializeRel(r)] = true
+	}
+
```

Comment:
What would you say about replacing this with something more like: 
```go
	allRelationships := map[Relationship]interface{}{}

	// index current list of relationships to ensure no duplication
	for _, r := range d.Relationships {
		allRelationships[*r] = nil
	}
```

Reply:
Ah you may be right about that. Unfortunate! We could refactor the relationship to be a relationship and a relationship with comment, e.g.:
```go
type BasicRelationship struct {
	RefA         common.DocElementID `json:"spdxElementId"`
	RefB         common.DocElementID `json:"relatedSpdxElement"`
	Relationship string
}

type Relationship struct {
	BasicRelationship

	RelationshipComment string `json:"comment,omitempty"`
}
```
... in which case it would be `allRelationships := map[BasicRelationship]struct{}{}`

Reply:
I don't think this would change any API surface area for users, they would just continue to reference Relationship

Reply:
I think this would change the way they create Relationships though - i think golang doesn't allow you to declare it embedded? but i may be wrong there

Reply:
If you have an embedded struct like:
```go
type Relationship struct {
	BasicRelationship
}
```
The fields (and methods) from BasicRelationship get "promoted" to the top-level  Relationship struct, e.g. I can write:
```go
r := Relationship{}
r.RefA = <something>
```
... and everything works as expected. Additionally, the go JSON library treats embedded fields like this without nesting them in the JSON structure.

Reply:
This definitely won't work as I had thought, so belay all talk of this embedded struct idea, sorry!